source/slang/slang-dataflow.h

```diff
@@ -0,0 +1,148 @@
+#pragma once
+
+#include "../core/slang-basic.h"
+
+#include <algorithm>
+#include <limits>
+#include <queue>
+
+namespace Slang
+{
+
+template<typename D>
+struct InOutPair
+{
+    D in;
+    D out;
+};
+
+//
+// Parameters:
+// - context: This object provides the transfer function
+// - ns: The list of nodes for which we should return the eventual output of
+//       their transfer function. The closure of predecessor nodes must be
+//       finite. Should not contain duplicates.
+// - strictlyDependsOn:
+//      A function f(x, y) which returns true if x depends on the output of y
+//      and y does not depend on x (and thus y should be evaluated before x)
+// - maxNumUpdates:
+//      (optional) the upper limit for how many times to evaluate the
+//      transfer function at each node.
+//
+// Returns:
+// - A list of eventual inputs and outputs of the transfer functions of the
+//   nodes specified in `ns`
+//
+template<typename T, typename P>
+List<InOutPair<typename T::Domain>> dataFlow(
+    T& context,
+    List<typename T::Node> ns,
+    P strictlyDependsOn,
+    typename T::Domain initialInput = T::Domain::bottom(),
+    UInt maxNumUpdates = std::numeric_limits<UInt>::max())
```

Comment:
`kMaxUInt`, the rest of the codebase uses std very sparingly, so let's keep it that way.